File: LargestRectangleInHistogram.py
def largestRectangleArea(height):
    stack = []
    maxarea = 0
    i = 0
    while i < len(height):
        if len(stack) == 0:
            stack.append(i)
            i += 1
            continue
        top = stack[-1]
        if height[i] >= height[top]:
            stack.append(i)
            i += 1
        else:
            stack.pop()
            width = i - stack[-1] - 1 if len(stack) > 0 else i
            area = width * height[top]
            if area > maxarea:
                maxarea = area
    print("after stack: " + str(maxarea))
    print("current stack len: " + str(len(stack)))
    while len(stack) > 0:
        top = stack.pop()
        width = len(height) - stack[-1] - 1 if len(stack) > 0 else len(height)
        area = width * height[top]
        if area > maxarea:
            maxarea = area
    return maxarea


if __name__ == '__main__':
    print(largestRectangleArea([5, 4, 1, 2]))

File: LargestRectangleInHistogram.cc
class Solution {
    public:
        int largestRectangleArea(vector<int> &height) {
            if (height.empty()) {
                return 0;
            }

            stack<int> maxes;
            int largest = INT_MIN;
            int i = 0;
            while(i < height.size()) {
                if (maxes.empty() || height[i] >= height[maxes.top()]) {
                    maxes.push(i);
                    i++;
                } else {
                    int curr_top_index = maxes.top();
                    maxes.pop();
                    int area;
                    if (maxes.empty()) {
                        area = height[maxes.top()] * i;
                    } else {
                        area = height[curr_top_index] * (i - 1 - maxes.top());
                    }
                    if (area > largest) {
                        largest = area;
                    }
                }
            }

            while(!maxes.empty()) {
                int curr_top_index = maxes.top();
                maxes.pop();
                int area;
                if (maxes.empty()) {
                    area = height[curr_top_index] * height.size();
                } else {
                    area = height[curr_top_index] * (height.size() - maxes.top() - 1);
                }
                if (area > largest) {
                    largest = area;
                }
            }

            return largest;
        }
};
